packages/dx-react-grid/src/plugins/grouping-state.jsx

```diff
@@ -4,106 +4,135 @@ import { Getter, Action, PluginContainer } from '@devexpress/dx-react-core';
+    this.draftGroupingChange = this.applyReducer.bind(this, draftGroupingChange);
+    this.cancelGroupingChange = this.applyReducer.bind(this, cancelGroupingChange);
+    this.setColumnSorting = this.setColumnSorting.bind(this);
+  }
+  getState() {
+    return {
+      ...this.state,
+      grouping: this.props.grouping || this.state.grouping,
+      expandedGroups: this.props.expandedGroups || this.state.expandedGroups,
     };
-
-    this.cancelGroupingChange = () => {
-      this.setState({
-        groupingChange: cancelGroupingChange(),
+  }
+  setColumnSorting({ columnName, keepOther, ...restParams }, { sorting }, { setColumnSorting }) {
+    const { grouping } = this.getState();
+    const groupingIndex = grouping
+      .findIndex(columnGrouping => columnGrouping.columnName === columnName);
+    if (groupingIndex === -1) {
+      setColumnSorting({
+        columnName,
+        keepOther: keepOther || grouping.map(columnGrouping => columnGrouping.columnName),
+        ...restParams,
       });
-    };
+      return false;
+    }
+
+    const sortIndex = Math.min(grouping
+      .reduce(
+        (acc, columnGrouping) =>
+          (sorting.findIndex(columnSorting =>
+            columnSorting.columnName === columnGrouping.columnName) === -1
+            ? acc
+            : acc + 1),
+        0), groupingIndex);
+    setColumnSorting({
+      columnName,
+      keepOther: true,
+      sortIndex,
+      ...restParams,
+    });
+    return false;
+  }
+  groupByColumn({ columnName, groupIndex }, { sorting }, { setColumnSorting }) {
+    const { grouping } = this.state;
+    const { grouping: newGrouping } = this.applyReducer(groupByColumn, { columnName, groupIndex });
+
+    if (!sorting) return;
+
+    const columnSortingIndex = sorting
+      .findIndex(columnSorting => columnSorting.columnName === columnName);
+    const groupingIndex = grouping
+      .findIndex(columnGrouping => columnGrouping.columnName === columnName);
+    const newGroupingIndex = newGrouping
+      .findIndex(columnGrouping => columnGrouping.columnName === columnName);
+
+    if (columnSortingIndex === -1
+      || (groupingIndex === grouping.length - 1 && newGroupingIndex === -1)) return;
+
+    const sortIndex = Math.min(newGrouping
+      .reduce(
+        (acc, columnGrouping) =>
+          (sorting.findIndex(columnSorting =>
+            columnSorting.columnName === columnGrouping.columnName) === -1
+            ? acc
+            : acc + 1),
+        0), newGroupingIndex === -1 ? newGrouping.length : newGroupingIndex);
+
+    if (columnSortingIndex === sortIndex) return;
+
+    setColumnSorting({
+      keepOther: true,
+      sortIndex,
+      ...sorting[columnSortingIndex],
+    });
+  }
+  applyReducer(reduce, payload) {
+    const state = this.getState();
+    const nextState = reduce(state, payload);
+    this.setState(nextState);
+
+    const { onGroupingChange } = this.props;
+    if (onGroupingChange && nextState.grouping !== state.grouping) {
```

Comment:
`nextState.grouping`, `nextState.expandedGroups` can be replaced to `const { grouping, expandedGroups } = nextState; `